tidy3d/components/geometry/base.py

```diff
@@ -2011,6 +2161,344 @@ def _surface_area(self, bounds: Bound) -> float:
+        z : float
+            Translation along z.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        return np.array(
+            [
+                (1.0, 0.0, 0.0, x),
+                (0.0, 1.0, 0.0, y),
+                (0.0, 0.0, 1.0, z),
+                (0.0, 0.0, 0.0, 1.0),
+            ],
+            dtype=float,
+        )
+
+    @staticmethod
+    def scaling(x: float, y: float = None, z: float = None) -> MatrixReal4x4:
+        """Return a scaling matrix.
+
+        Parameters
+        ----------
+        x : float
+            Scaling factor along x.
+        y : float = None
+            Scaling factor along y. If ``None``, use the value for x.
+        z : float = None
+            Scaling factor along z. If ``None``, use the value for x.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        if y is None:
+            y = x
+
+        if z is None:
+            z = x
+
+        if np.isclose((x, y, z), 0.0).any():
+            raise Tidy3dError("Scaling factors cannot be zero in any dimensions.")
+        return np.array(
+            [
+                (x, 0.0, 0.0, 0.0),
+                (0.0, y, 0.0, 0.0),
+                (0.0, 0.0, z, 0.0),
+                (0.0, 0.0, 0.0, 1.0),
+            ],
+            dtype=float,
+        )
+
+    @staticmethod
+    def rotation(angle: float, axis: Union[Axis, Coordinate]) -> MatrixReal4x4:
+        """Return a rotation matrix.
+
+        Parameters
+        ----------
+        angle : float
+            Rotation angle (in radians).
+        axis : Union[int, Tuple[float, float, float]]
+            Axis of rotation: 0, 1, or 2 for x, y, and z, respectively, or a 3D vector.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        sin = np.sin(angle)
+        cos = np.cos(angle)
+        if axis == 0:
+            return np.array(
+                [
+                    (1.0, 0.0, 0.0, 0.0),
+                    (0.0, cos, -sin, 0.0),
+                    (0.0, sin, cos, 0.0),
+                    (0.0, 0.0, 0.0, 1.0),
+                ]
+            )
+
+        if axis == 1:
+            # NOTE (Lucas): Usual rotation transposed because of our axis ordering
+            return np.array(
+                [
+                    (cos, 0.0, -sin, 0.0),
+                    (0.0, 1.0, 0.0, 0.0),
+                    (sin, 0.0, cos, 0.0),
+                    (0.0, 0.0, 0.0, 1.0),
+                ]
+            )
+
+        if axis == 2:
+            return np.array(
+                [
+                    (cos, -sin, 0.0, 0.0),
+                    (sin, cos, 0.0, 0.0),
+                    (0.0, 0.0, 1.0, 0.0),
+                    (0.0, 0.0, 0.0, 1.0),
+                ]
+            )
+
+        try:
+            x, y, z = axis
```

Comment:
related to above comment, I guess we should also try to ensure that the shape of axis is (3,) or I guess (N,3) or (3, N) if we want to support vectorizing. I feel there could be some possible edge case errors that might come  in here that we can nip in the bud.

Reply:
I see. I've included a type "normalization" before any comparisons, which should take care of any valid inputs. I'm not allowing vectorization at this point, so I'm treating that case as an error.

Reply:
sounds good.